### PR TITLE
Fix Factory schema

### PR DESCRIPTION
**Pull request by cassava**
This change is [Reviewable](https://reviewable.io/reviews/eclipse/cloe/2)

### CODE REVIEW

File: fable/include/fable/schema/interface.hpp [hunk 1]
```diff
@@ -242,7 +247,13 @@ class Box : public Interface {
   Box(Interface* i) : impl_(i) { assert(impl_); }                             // NOLINT
   Box(std::shared_ptr<Interface> i) : impl_(std::move(i)) { assert(impl_); }  // NOLINT
 
- public:
+ public:  // Special
+  template <typename T>
+  std::shared_ptr<const T> as() const {
```
Comment (cassava) [inline fable/include/fable/schema/interface.hpp:252]: - [ ] Unit test this.

File: fable/include/fable/schema/struct.hpp [hunk 2]
```diff
@@ -111,11 +170,18 @@ class Struct : public Base<Struct> {
     return std::move(*this);
   }
 
-  template <typename T, std::enable_if_t<std::is_base_of<Interface, T>::value, int> = 0>
-  Struct additional_properties(const T& s) && {
+  // XXX(ben): This does not work!
```
Comment (cassava) [inline fable/include/fable/schema/struct.hpp:173]: - [ ] Fix this.

File: fable/include/fable/schema/struct.hpp [hunk 3]
```diff
@@ -111,11 +170,18 @@ class Struct : public Base<Struct> {
...
+  template <typename S, typename = enable_if_schema_t<S>>
+  void set_additional_properties(const S& s) {
     additional_properties_ = true;
-    additional_prototype_ = s.clone();
+    additional_prototype_.reset(s.clone());
     additional_prototype_->reset_ptr();
+  }
+
+  // XXX(ben): This does not work!
+  template <typename S, typename = enable_if_schema_t<S>>
```
Comment (cassava) [inline fable/include/fable/schema/struct.hpp:182]: - [ ] Fix `additional_properties()`

File: fable/include/fable/schema/factory.hpp [hunk 4]
```diff
@@ -43,36 +43,185 @@ namespace fable {
...
+    return std::move(*dynamic_cast<CRTP*>(this));
+  }
+
+  /**
+   * Set the args key and return this for chaining.
+   *
+   * \see set_args_key()
+   */
+  CRTP args_key(const std::string& keyword) && {
+    set_args_key(keyword);
```
Comment (cassava) [inline fable/include/fable/schema/factory.hpp:111]: - [ ] Test that no segfault occurs with `FactoryBase::set_args_key()`

**Review by cassava — COMMENTED** [review]
> Reviewed 3 of 3 files at r1, 2 of 2 files at r2, 3 of 3 files at r3, 19 of 21 files at r4, 4 of 8 files at r6, 5 of 5 files at r7.
> *[Reviewable](https://reviewable.io/reviews/eclipse/cloe/2)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by cassava — COMMENTED** [review]
> Reviewed 1 of 6 files at r8, 5 of 5 files at r9.
> *[Reviewable](https://reviewable.io/reviews/eclipse/cloe/2)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @clssn and @tobifalk)

**Review by tobifalk — COMMENTED** [review]
> Reviewed 2 of 3 files at r1, 2 of 3 files at r3, 5 of 21 files at r4, 3 of 8 files at r6, 1 of 5 files at r9.
> *[Reviewable](https://reviewable.io/reviews/eclipse/cloe/2)* status: all files reviewed, 7 unresolved discussions (waiting on @cassava and @clssn)

___
*[fable/include/fable/utility/gtest.hpp, line 84 at r9](https://reviewable.io/reviews/eclipse/cloe/2#-MTFMon546lfuRqyD_1G:-MTFMon546lfuRqyD_1H:b-jkf4w6) ([raw file](https://github.com/eclipse/cloe/blob/05985ac9ed443deb4380d9949e45fcbb49aad71b/fable/include/fable/utility/gtest.hpp#L84)):*
> ```cpp
> }
> 
> inline void assert_from_eq_to(Confable& x, const Json& identity) {
> ```

If the first parameter is of type `Confable` for all variants of this function, `assert_from_conf_eq_to` might be more readable.
___
*[fable/src/fable/schema_test.cpp, line 150 at r9](https://reviewable.io/reviews/eclipse/cloe/2#-MTFhq99Bw04VJVjpaWP:-MTFhq99Bw04VJVjpaWQ:bo4yndg) ([raw file](https://github.com/eclipse/cloe/blob/05985ac9ed443deb4380d9949e45fcbb49aad71b/fable/src/fable/schema_test.cpp#L150)):*
> ```cpp
>   MyStruct s2;
> 
>   ASSERT_EQ(s1.json_schema().dump(), s2.schema().json_schema().dump());
> ```

`fable::assert_eq(s1.json_schema(), s2.schema().json_schema());` ?
___
*[fable/src/fable/schema/enum_test.cpp, line 79 at r9](https://reviewable.io/reviews/eclipse/cloe/2#-MTFVk4g5do92Kss1EPO:-MTFVk4g5do92Kss1EPP:bvixrew) ([raw file](https://github.com/eclipse/cloe/blob/05985ac9ed443deb4380d9949e45fcbb49aad71b/fable/src/fable/schema/enum_test.cpp#L79)):*
> ```cpp
>   fable::Json j;
>   s.to_json(j);
>   ASSERT_EQ(j.dump(2), empty.dump(2));
> ```

Why not use `fable::assert_eq(j, empty);` here?
___
*[fable/src/fable/schema/enum_test.cpp, line 86 at r9](https://reviewable.io/reviews/eclipse/cloe/2#-MTFWH8p4FnqHzcl3alZ:-MTFWH8p4FnqHzcl3al_:b-s9y44a) ([raw file](https://github.com/eclipse/cloe/blob/05985ac9ed443deb4380d9949e45fcbb49aad71b/fable/src/fable/schema/enum_test.cpp#L86)):*
> ```cpp
>   ASSERT_NO_THROW({
>     s.to_json(j);
>     ASSERT_EQ(j.dump(2), filled.dump(2));
> ```

How about `fable::assert_eq(j, filled);`?
___
*[fable/src/fable/schema/enum_test.cpp, line 105 at r9](https://reviewable.io/reviews/eclipse/cloe/2#-MTFX2fXEkO2RyTIYCXO:-MTFX2fXEkO2RyTIYCXP:b-d5exl5) ([raw file](https://github.com/eclipse/cloe/blob/05985ac9ed443deb4380d9949e45fcbb49aad71b/fable/src/fable/schema/enum_test.cpp#L105)):*
> ```cpp
>   fable::Json j;
>   s.to_json(j);
>   ASSERT_EQ(j.dump(2), empty.dump(2));
> ```

`fable::assert_eq(j, empty);`
___
*[fable/src/fable/schema/enum_test.cpp, line 110 at r9](https://reviewable.io/reviews/eclipse/cloe/2#-MTFXCw-6OEv0HRHpTcY:-MTFXCw-6OEv0HRHpTcZ:b-fvfncb) ([raw file](https://github.com/eclipse/cloe/blob/05985ac9ed443deb4380d9949e45fcbb49aad71b/fable/src/fable/schema/enum_test.cpp#L110)):*
> ```cpp
>   ASSERT_EQ(xs.size(), 1);
>   ASSERT_EQ(*(xs[0].level), LogLevel::info);
>   ASSERT_EQ(fable::Json(xs).dump(2), filled.dump(2));
> ```

`fable::assert_eq(xs, filled);`
___
*[fable/src/fable/schema/enum_test.cpp, line 113 at r9](https://reviewable.io/reviews/eclipse/cloe/2#-MTFXHei0H5BCW_7L5Za:-MTFXHei0H5BCW_7L5Zb:bnnmxp2) ([raw file](https://github.com/eclipse/cloe/blob/05985ac9ed443deb4380d9949e45fcbb49aad71b/fable/src/fable/schema/enum_test.cpp#L113)):*
> ```cpp
>   ASSERT_NO_THROW({
>     s.to_json(j);
>     ASSERT_EQ(j.dump(2), filled.dump(2));
> ```

`fable::assert_eq(j, filled);`

File: fable/src/fable/schema/struct_test.cpp [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+};
+
+struct StructSub1Sub1 : public StructSub1 {
+  int subsub1 = 2;
+};
+
+}  // anonymous namespace
+
+TEST(fable_schema_struct, base_schema) {
+  StructFoo tmp;
```
Comment (clssn) [inline fable/src/fable/schema/struct_test.cpp:221]: You want to test StructBase, StructSub1, StructSub2 and StructSubSub1.

**Review by clssn — CHANGES_REQUESTED** [review]

**Review by cassava — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/eclipse/cloe/2)* status: all files reviewed, 6 unresolved discussions (waiting on @tobifalk)

___
*[fable/include/fable/utility/gtest.hpp, line 84 at r9](https://reviewable.io/reviews/eclipse/cloe/2#-MTFMon546lfuRqyD_1G:-MTdi_ZM0pMGzVST3F8N:b-eap5nh) ([raw file](https://github.com/eclipse/cloe/blob/05985ac9ed443deb4380d9949e45fcbb49aad71b/fable/include/fable/utility/gtest.hpp#L84)):*
<details><summary><i>Previously, tobifalk (Tobias Falkenstein) wrote…</i></summary><blockquote>

If the first parameter is of type `Confable` for all variants of this function, `assert_from_conf_eq_to` might be more readable.
</blockquote></details>

The `from_conf` part refers to the second argument, not the first. With the `assert_from_conf` function, we actually convert the second argument to `Conf`, which we could also just use. That is not the case with this function.
___
*[fable/src/fable/schema_test.cpp, line 150 at r9](https://reviewable.io/reviews/eclipse/cloe/2#-MTFhq99Bw04VJVjpaWP:-MTdqsvG6vzFL_5_kcn2:b-896fix) ([raw file](https://github.com/eclipse/cloe/blob/05985ac9ed443deb4380d9949e45fcbb49aad71b/fable/src/fable/schema_test.cpp#L150)):*
<details><summary><i>Previously, tobifalk (Tobias Falkenstein) wrote…</i></summary><blockquote>

`fable::assert_eq(s1.json_schema(), s2.schema().json_schema());` ?
</blockquote></details>

Done.
___
*[fable/src/fable/schema/enum_test.cpp, line 79 at r9](https://reviewable.io/reviews/eclipse/cloe/2#-MTFVk4g5do92Kss1EPO:-MTe724x6ul53qkWHh79:b-896fix) ([raw file](https://github.com/eclipse/cloe/blob/05985ac9ed443deb4380d9949e45fcbb49aad71b/fable/src/fable/schema/enum_test.cpp#L79)):*
<details><summary><i>Previously, tobifalk (Tobias Falkenstein) wrote…</i></summary><blockquote>

Why not use `fable::assert_eq(j, empty);` here?
</blockquote></details>

Done.
___
*[fable/src/fable/schema/enum_test.cpp, line 86 at r9](https://reviewable.io/reviews/eclipse/cloe/2#-MTFWH8p4FnqHzcl3alZ:-MTe72tp3ojxrCTqoI7Y:b-896fix) ([raw file](https://github.com/eclipse/cloe/blob/05985ac9ed443deb4380d9949e45fcbb49aad71b/fable/src/fable/schema/enum_test.cpp#L86)):*
<details><summary><i>Previously, tobifalk (Tobias Falkenstein) wrote…</i></summary><blockquote>

How about `fable::assert_eq(j, filled);`?
</blockquote></details>

Done.
___
*[fable/src/fable/schema/enum_test.cpp, line 105 at r9](https://reviewable.io/reviews/eclipse/cloe/2#-MTFX2fXEkO2RyTIYCXO:-MTe74061c2Ld9S8gegy:b-896fix) ([raw file](https://github.com/eclipse/cloe/blob/05985ac9ed443deb4380d9949e45fcbb49aad71b/fable/src/fable/schema/enum_test.cpp#L105)):*
<details><summary><i>Previously, tobifalk (Tobias Falkenstein) wrote…</i></summary><blockquote>

`fable::assert_eq(j, empty);`
</blockquote></details>

Done.
___
*[fable/src/fable/schema/enum_test.cpp, line 110 at r9](https://reviewable.io/reviews/eclipse/cloe/2#-MTFXCw-6OEv0HRHpTcY:-MTe74j250sVHTXDAOdJ:b-896fix) ([raw file](https://github.com/eclipse/cloe/blob/05985ac9ed443deb4380d9949e45fcbb49aad71b/fable/src/fable/schema/enum_test.cpp#L110)):*
<details><summary><i>Previously, tobifalk (Tobias Falkenstein) wrote…</i></summary><blockquote>

`fable::assert_eq(xs, filled);`
</blockquote></details>

Done.
___
*[fable/src/fable/schema/enum_test.cpp, line 113 at r9](https://reviewable.io/reviews/eclipse/cloe/2#-MTFXHei0H5BCW_7L5Za:-MTe7571447v3bLasI0e:b-896fix) ([raw file](https://github.com/eclipse/cloe/blob/05985ac9ed443deb4380d9949e45fcbb49aad71b/fable/src/fable/schema/enum_test.cpp#L113)):*
<details><summary><i>Previously, tobifalk (Tobias Falkenstein) wrote…</i></summary><blockquote>

`fable::assert_eq(j, filled);`
</blockquote></details>

Done.

File: fable/include/fable/schema/factory.hpp [hunk 6]
```diff
@@ -91,55 +240,78 @@ class Factory : public Base<Factory<T, F>> {
...
-          {"name", String{nullptr, "identifier of simulator instance"}},
-          {"args", kv.second->schema()},
-        }.additional_properties(true).reset_pointer()
-      );
-      // clang-format on
+      Struct base{
+          {factory_key_, make_const_str(kv.first, "name of factory").require()},
+      };
+      if (args_key_ == "") {
+        // FIXME(ben): This branch leads to seg-faults in production code. Not
```
Comment (clssn) [inline fable/include/fable/schema/factory.hpp:303]: Test this on noisy_sensor.

File: fable/include/fable/schema/factory.hpp [hunk 7]
```diff
@@ -154,9 +326,75 @@ class Factory : public Base<Factory<T, F>> {
     return out;
   }
 
- private:
+ protected:
   std::shared_ptr<Variant> schema_;
-  std::map<std::string, std::shared_ptr<F>> available_;
+  TransformFunc transform_func_;
+  std::vector<Box> available_schemas_;
```
Comment (clssn) [inline fable/include/fable/schema/factory.hpp:333]: Unused. Delete this.

File: fable/src/fable/schema/factory_test.cpp [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+                ],
+                "type": "object"
+              },
+              "factory": {
+                "const": "prime",
+                "description": "name of factory"
+              }
+            },
+            "required": [
+              "factory"
```
Comment (clssn) [inline fable/src/fable/schema/factory_test.cpp:124]: What if a factory wants args? --> New issue.

Comment (cassava) [reply fable/src/fable/schema/factory_test.cpp:124]: Done: #23

File: engine/src/stack.hpp [hunk 9]
```diff
@@ -683,67 +716,50 @@ struct VehicleConf : public Confable {
...
-    ptr_->to_json(j);
+  void to_json(Json&) const override {
+    throw std::logic_error("VehicleSchema does not implement to_json");
   }
 
-  void reset_ptr() override { ptr_ = nullptr; }
+  void reset_ptr() override {}
 
  private:  // State
   ComponentSchema components_;
```
[the remark this answers is not in the document]
Comment (cassava) [reply engine/src/stack.hpp:765]: Good catch.

File: engine/src/stack.cpp [hunk 10]
```diff
@@ -101,6 +101,31 @@ std::string PluginConf::canonical() const {
...
+SimulatorSchema::SimulatorSchema() {
+  this->set_transform_schema([](fable::schema::Struct&& s) -> fable::schema::Box {
+    s.set_property("name", id_prototype("globally unique identifier for simulator"));
+    return s;
+  });
+}
+
+ControllerSchema::ControllerSchema() {
+  this->set_transform_schema([](fable::schema::Struct&& s) -> fable::schema::Box {
+    s.set_property("name", id_prototype("globally unique identifier for controller"));
```
Comment (tobifalk) [inline engine/src/stack.cpp:113]: Now, there is some redundancy between the factory schema transformation defined here and the schema definition in `ControllerConf` (s. [here](https://github.com/eclipse/cloe/blob/ben/fix-factory-schema/engine/src/stack.hpp#L549)). But I guess a solution is out of the scope of this PR

Comment (cassava) [reply engine/src/stack.cpp:113]: That is correct, and is also out of the scope of this PR. One step at a time, in particular once the benefit is high enough.

**Review by clssn — APPROVED** [review]
LGTM!

**Review by tobifalk — APPROVED** [review]
Looks good!

**Pull request merged**